add script to check for migration order

roperzh opened the pull request:
Related to #6142, this adds a CI check for the order of migrations.

As I noted in a comment on the workflow file, it's important to keep in mind that some migrations might still go unnoticed even with this check, example:

1. PR1 adds a migration, CI check pass
2. PR2 adds a migration, CI pass, gets merged
3. PR1 can still be merged because the CI checks aren't run again

The check will fail in `main` however, so if we find the current script to be reliable, we could setup a Slack ping or something similar, to make sure somebody takes a look

---

This is how a [failed test](https://github.com/fleetdm/fleet/actions/runs/3100710308/jobs/5021283396) looks like:

<img width="1021" alt="image" src="https://user-images.githubusercontent.com/4419992/191598936-2c07f227-e9aa-4b15-acce-64930621059e.png">


This is how a [successful test](https://github.com/fleetdm/fleet/actions/runs/3101151978/jobs/5022207471) looks like

<img width="1069" alt="image" src="https://user-images.githubusercontent.com/4419992/191611969-5c032ed4-ec56-4dda-8e2f-7cf0145b6a20.png">

michalnicp commented on .github/workflows/test-db-changes.yml:
```diff
@@ -0,0 +1,61 @@
+    - name: Check migration order
+      run: |
+        all_migrations=($(ls -1 server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
+        new_migrations=($(git diff --name-only origin/main -- server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
```
Also, I ran into some issues with this script when I was testing. See https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511

```
Run all_migrations=($(ls -1 server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
[2](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:2)
  all_migrations=($(ls -1 server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
[3](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:3)
  new_migrations=($(git diff --name-only origin/main -- server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
[4](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:4)
  
[5](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:5)
  index=0
[6](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:6)
  for migration in $new_migrations; do
[7](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:7)
    if [ "$migration" != "${all_migrations[$index]}" ]; then
[8](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:8)
      echo "❌ fail: $migration has an older timestamp than ${all_migrations[$index]}"
[9](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:9)
      echo "this might cause problems if this change is merged"
[10](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:10)
      echo "please update the timestamp of $migration"
[11](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:11)
      exit 1
[12](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:12)
    fi
[13](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:13)
    ((index++))
[14](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:14)
  done
[15](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:15)
  shell: /usr/bin/bash -e {0}
[16](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:16)
  
[16](https://github.com/fleetdm/fleet/actions/runs/3070823413/jobs/4960937511#step:6:17)
fatal: bad revision 'origin/main'
```

roperzh replied:
thanks so much for taking the time to test this. It was working but I broke it in a cleanup commit. I will fix and test again.

roperzh commented on .github/workflows/test-db-changes.yml:
```diff
@@ -0,0 +1,61 @@
+    # when this check fails on `main`.
+    - name: Check migration order
+      run: |
+        all_migrations=($(ls -1 server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
```
TIL, thanks I will remove

roperzh commented:
@michalnicp thanks so much again for testing, I have fixed the issues and adjusted with your suggestions.

I have also re-tested everything (ref: the latest three commits on this branch) and updated the issue description.

Would you be so kind to take another look?

chiiph commented on .github/workflows/test-db-changes.yml:
```diff
@@ -0,0 +1,62 @@
+    - name: Check migration order
+      run: |
+        all_migrations=($(ls server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
+        new_migrations=($(git diff --name-only origin/${{github.base_ref}} -- server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
```
Why will this work in main in the example you mentioned? Would `github.base_ref` be one of the two PRs somehow?

roperzh replied:
yes, after `actions/checkout` does the git checkout, the current HEAD is what they call the "merge commit" i.e. github.base_ref + the PR changes merged.

So when we run `git diff origin/github.base_ref`, the diff lists only the files modified by the PR.

chiiph replied:
So if this runs on main at any random point, it won't work, it has to run when the merge commit hits main... it won't be caught in our nightly builds unless this happens to be the last commit in main at the time. Am I understanding correctly?

I think this is fine, it'll help us catch more cases, but it's not bullet proof. Not sure if anything beats the test that deploys version A, then migrates to version B and runs fleet serve. But that's a hard one to automate.

chiiph replied:
> We can modify the workflow to use the latest released tag when it's run on main instead of github.base_ref, and I think that will fix this particular issue, right?

I don't think I understand.

> Having said that, if we think that actually doing the migration is a net positive (it might even also catch other cases?) I don't think it's very difficult to implement honestly! (this workflow already has a DB instance running!)

It wouldn't be just running the migration, it would be also starting fleet serve to make sure it runs its usual checks.

The way it would work is:

1. Checkout latest tag and build fleet
2. Run migrations
3. Start fleet serve
4. Wait a bit, if fleet exits on its own, FAIL, otherwise keep going
5. Kill fleet serve
6. Checkout main and build fleet
7. Run migrations
8. Start fleet serve
9. Wait a bit, if fleet exits on its own, FAIL, otherwise SUCCESS

Ideally, we would test the past 2 or 3 releases.

roperzh replied:
> I don't think I understand.

I was thinking on something like this pseudo-code:

```
if base_branch == "main":
  comp = get_latest_tag()
else
  comp = github.base_ref

new_migrations = $("git diff --name-only $comp -- ...")
```

so, if the workflow runs in main, we compare the migrations against the latest released fleet version (or last two, etc). This should cover the case you raised if I understand correctly

> It wouldn't be just running the migration, it would be also starting fleet serve to make sure it runs its usual checks.

yeah, I expressed myself poorly, but I had in mind something akin to what you described (thanks btw!)

I was planning on merging this _after_ the release, so we don't have any surprises during the freeze week, I will see how hard is to do the approach you described before then.

chiiph replied:
> I was thinking on something like this pseudo-code:

Ah, yeah, that could work. Worth giving it a try.

roperzh replied:
@chiiph I updated a branch with what we discussed here, would you mind taking another look? 🙏
